src/code/z_actor.c

```diff
@@ -3365,110 +3407,156 @@ s32 func_800BB59C(PlayState* play, Actor* actor) {
-    f32 temp_f0_2;
-    Actor* targetedActor;
-    Actor* actor;
-    s32 phi_s2;
-    CollisionPoly* sp80;
-    s32 sp7C;
-    Vec3f sp70;
+/**
+ * Search for targetable actors of the `actorCategory` category.
+ *
+ * Looks for the actor of said category with higher targetPriority and the one that is nearest to player. This actor
+ * must be within the range (relative to player) speicified by its targetMode.
+ *
+ * The actor must be on-screen (more or less, not sure what's going on in Target_InTargetableScreenRegion)
+ *
+ * The highest priority actor is stored in `sTargetableHighestPriorityActor`, while the nearest actor is stored in
+ * `sTargetableNearestActor`. The higher priority / smaller distance of those actors are stored in
+ * `sTargetableHighestPriorityPriority` and `sTargetableNearestActorDistanceSq`.
+ *
+ * Something something, ACTOR_FLAG_40000000, D_801ED8C4/D_801ED8C0, D_801ED8D8/D_801ED8D0.
+ *
+ * For an actor to be taken in consideration by this function it needs:
+ * - Non-NULL update function (maybe obvious?)
+ * - Not be Player itself.
+ * - It must be targetable or ACTOR_FLAG_40000000
+ * - Not be the already targeted actor, unless it has the ACTOR_FLAG_80000 flag
+ * - Be withing the range specified by its targetMode.
+ * - It must be on-screen (within a margin)
+ * - Must not be blocked by a surface (?)
+ *
+ * This function also checks for the nearest enemy actor, which allows determining if enemy background music should be
+ * played. This actor is stored in `targetContext.bgmEnemy` and its distance is stored in `sBgmEnemyDistSq`
+ *
+ * This function is expected to be called with almost every actor category in each cycle. On a new cycle its global
+ * variables must be reset by the caller, otherwise the information of the previous cycle will be retained on this one.
+ */
+void Target_FindTargetableActorForCategory(PlayState* play, ActorContext* actorCtx, Player* player,
+                                           ActorType actorCategory) {
+    f32 distSq;
+    Actor* actor = actorCtx->actorLists[actorCategory].first;
+    Actor* targetedActor = player->targetedActor;
+    s32 isNearestTargetableActor;
     s32 phi_s2_2;
 
-    actor = actorCtx->actorLists[actorCategory].first;
-    targetedActor = player->targetedActor;
-    while (actor != NULL) {
-        if ((actor->update != NULL) && ((Player*)actor != player)) {
-            if (actor->flags & (ACTOR_FLAG_40000000 | ACTOR_FLAG_1)) {
-                if ((actorCategory == ACTORCAT_ENEMY) && CHECK_FLAG_ALL(actor->flags, ACTOR_FLAG_4 | ACTOR_FLAG_1)) {
-                    if ((actor->xyzDistToPlayerSq < SQ(500.0f)) && (actor->xyzDistToPlayerSq < sBgmEnemyDistSq)) {
-                        actorCtx->targetContext.bgmEnemy = actor;
-                        sBgmEnemyDistSq = actor->xyzDistToPlayerSq;
-                    }
-                }
+    for (; actor != NULL; actor = actor->next) {
+        if ((actor->update == NULL) || ((Player*)actor == player)) {
```

Comment:
```suggestion
        if ((actor->update == NULL) || (actor == &player->actor)) {
```
This probably makes more sense?

Reply:
Sadly doesn't match, was one of the very first stuff I tried